// tentaclaw-gateway/dist/gateway/src/db.js
"use strict";
/**
 * TentaCLAW Gateway — Database Layer (SQLite)
 *
 * Self-hosted. No SaaS. Your data stays on your hardware.
 * CLAWtopus says: "I remember everything. Eight arms, one brain."
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDb = getDb;
exports.registerNode = registerNode;
exports.getNode = getNode;
exports.getAllNodes = getAllNodes;
exports.getNodesByFarm = getNodesByFarm;
exports.deleteNode = deleteNode;
exports.updateNodeStatus = updateNodeStatus;
exports.markStaleNodes = markStaleNodes;
exports.insertStats = insertStats;
exports.getStatsHistory = getStatsHistory;
exports.pruneStats = pruneStats;
exports.queueCommand = queueCommand;
exports.getPendingCommands = getPendingCommands;
exports.completeCommand = completeCommand;
exports.createFlightSheet = createFlightSheet;
exports.getAllFlightSheets = getAllFlightSheets;
exports.getFlightSheet = getFlightSheet;
exports.deleteFlightSheet = deleteFlightSheet;
exports.applyFlightSheet = applyFlightSheet;
exports.createAlert = createAlert;
exports.getRecentAlerts = getRecentAlerts;
exports.acknowledgeAlert = acknowledgeAlert;
exports.checkAndAlert = checkAndAlert;
exports.storeBenchmark = storeBenchmark;
exports.getNodeBenchmarks = getNodeBenchmarks;
exports.getAllBenchmarks = getAllBenchmarks;
exports.getClusterSummary = getClusterSummary;
exports.recordNodeEvent = recordNodeEvent;
exports.getNodeEvents = getNodeEvents;
exports.getCompactHistory = getCompactHistory;
exports.findBestNode = findBestNode;
exports.getClusterModels = getClusterModels;
exports.getHealthScore = getHealthScore;
exports.createSchedule = createSchedule;
exports.getSchedule = getSchedule;
exports.getAllSchedules = getAllSchedules;
exports.deleteSchedule = deleteSchedule;
exports.toggleSchedule = toggleSchedule;
exports.markScheduleRun = markScheduleRun;
exports.getDueSchedules = getDueSchedules;
exports.addSshKey = addSshKey;
exports.getNodeSshKeys = getNodeSshKeys;
exports.deleteSshKey = deleteSshKey;
exports.addNodeTag = addNodeTag;
exports.removeNodeTag = removeNodeTag;
exports.getNodeTags = getNodeTags;
exports.getNodesByTag = getNodesByTag;
exports.getAllTags = getAllTags;
exports.startModelPull = startModelPull;
exports.updateModelPull = updateModelPull;
exports.getActiveModelPulls = getActiveModelPulls;
exports.getAllActiveModelPulls = getAllActiveModelPulls;
const better_sqlite3_1 = __importDefault(require("better-sqlite3"));
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
// =============================================================================
// Database Setup
// =============================================================================
const DB_PATH = process.env.TENTACLAW_DB_PATH || path_1.default.join(process.cwd(), 'data', 'hivemind.db');
let db;
function getDb() {
    if (!db) {
        const dir = path_1.default.dirname(DB_PATH);
        if (!fs_1.default.existsSync(dir)) {
            fs_1.default.mkdirSync(dir, { recursive: true });
        }
        db = new better_sqlite3_1.default(DB_PATH);
        db.pragma('journal_mode = WAL');
        db.pragma('foreign_keys = ON');
        initSchema(db);
    }
    return db;
}
function initSchema(db) {
    db.exec(`
        CREATE TABLE IF NOT EXISTS nodes (
            id TEXT PRIMARY KEY,
            farm_hash TEXT NOT NULL,
            hostname TEXT NOT NULL,
            ip_address TEXT,
            mac_address TEXT,
            registered_at TEXT DEFAULT (datetime('now')),
            last_seen_at TEXT,
            status TEXT DEFAULT 'online',
            gpu_count INTEGER DEFAULT 0,
            os_version TEXT
        );

        CREATE TABLE IF NOT EXISTS stats (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            node_id TEXT NOT NULL REFERENCES nodes(id) ON DELETE CASCADE,
            timestamp TEXT DEFAULT (datetime('now')),
            payload TEXT NOT NULL,
            gpu_count INTEGER,
            cpu_usage_pct REAL,
            ram_used_mb INTEGER,
            ram_total_mb INTEGER,
            toks_per_sec REAL
        );

        CREATE INDEX IF NOT EXISTS idx_stats_node_time ON stats(node_id, timestamp DESC);

        CREATE TABLE IF NOT EXISTS commands (
            id TEXT PRIMARY KEY,
            node_id TEXT NOT NULL REFERENCES nodes(id) ON DELETE CASCADE,
            action TEXT NOT NULL,
            payload TEXT,
            status TEXT DEFAULT 'pending',
            created_at TEXT DEFAULT (datetime('now')),
            sent_at TEXT,
            completed_at TEXT
        );

        CREATE INDEX IF NOT EXISTS idx_commands_node_status ON commands(node_id, status);

        CREATE TABLE IF NOT EXISTS flight_sheets (
            id TEXT PRIMARY KEY,
            name TEXT NOT NULL,
            description TEXT DEFAULT '',
            targets TEXT NOT NULL,
            created_at TEXT DEFAULT (datetime('now')),
            updated_at TEXT
        );

        CREATE TABLE IF NOT EXISTS alerts (
            id TEXT PRIMARY KEY,
            node_id TEXT,
            severity TEXT NOT NULL,
            type TEXT NOT NULL,
            message TEXT NOT NULL,
            value REAL,
            threshold REAL,
            acknowledged INTEGER DEFAULT 0,
            created_at TEXT DEFAULT (datetime('now'))
        );

        CREATE INDEX IF NOT EXISTS idx_alerts_node ON alerts(node_id, created_at DESC);

        CREATE TABLE IF NOT EXISTS benchmarks (
            id TEXT PRIMARY KEY,
            node_id TEXT NOT NULL REFERENCES nodes(id) ON DELETE CASCADE,
            model TEXT NOT NULL,
            tokens_per_sec REAL NOT NULL,
            prompt_eval_rate REAL DEFAULT 0,
            eval_rate REAL DEFAULT 0,
            total_duration_ms INTEGER DEFAULT 0,
            created_at TEXT DEFAULT (datetime('now'))
        );

        CREATE INDEX IF NOT EXISTS idx_benchmarks_node ON benchmarks(node_id, created_at DESC);

        CREATE TABLE IF NOT EXISTS node_events (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            node_id TEXT NOT NULL,
            event TEXT NOT NULL,
            detail TEXT,
            created_at TEXT DEFAULT (datetime('now'))
        );

        CREATE INDEX IF NOT EXISTS idx_node_events ON node_events(node_id, created_at DESC);

        CREATE TABLE IF NOT EXISTS schedules (
            id TEXT PRIMARY KEY,
            name TEXT NOT NULL,
            type TEXT NOT NULL,
            cron TEXT NOT NULL,
            config TEXT NOT NULL,
            enabled INTEGER DEFAULT 1,
            last_run TEXT,
            next_run TEXT,
            created_at TEXT DEFAULT (datetime('now'))
        );

        CREATE TABLE IF NOT EXISTS ssh_keys (
            id TEXT PRIMARY KEY,
            node_id TEXT NOT NULL REFERENCES nodes(id) ON DELETE CASCADE,
            label TEXT NOT NULL,
            public_key TEXT NOT NULL,
            fingerprint TEXT NOT NULL,
            created_at TEXT DEFAULT (datetime('now'))
        );

        CREATE INDEX IF NOT EXISTS idx_ssh_keys_node ON ssh_keys(node_id);

        CREATE TABLE IF NOT EXISTS node_tags (
            node_id TEXT NOT NULL REFERENCES nodes(id) ON DELETE CASCADE,
            tag TEXT NOT NULL,
            created_at TEXT DEFAULT (datetime('now')),
            PRIMARY KEY (node_id, tag)
        );

        CREATE INDEX IF NOT EXISTS idx_node_tags_tag ON node_tags(tag);

        CREATE TABLE IF NOT EXISTS model_pulls (
            id TEXT PRIMARY KEY,
            node_id TEXT NOT NULL REFERENCES nodes(id) ON DELETE CASCADE,
            model TEXT NOT NULL,
            status TEXT DEFAULT 'downloading',
            progress_pct REAL DEFAULT 0,
            bytes_downloaded INTEGER DEFAULT 0,
            bytes_total INTEGER DEFAULT 0,
            started_at TEXT DEFAULT (datetime('now')),
            updated_at TEXT DEFAULT (datetime('now'))
        );

        CREATE INDEX IF NOT EXISTS idx_model_pulls_node ON model_pulls(node_id);
    `);
}
// =============================================================================
// Helpers
// =============================================================================
function generateId() {
    return Date.now().toString(36) + Math.random().toString(36).slice(2, 8);
}
// =============================================================================
// Node Operations
// =============================================================================
function registerNode(reg) {
    const d = getDb();
    const existing = d.prepare('SELECT id FROM nodes WHERE id = ?').get(reg.node_id);
    if (existing) {
        d.prepare(`
            UPDATE nodes SET
                farm_hash = ?, hostname = ?, ip_address = ?, mac_address = ?,
                gpu_count = ?, os_version = ?, status = 'online', last_seen_at = datetime('now')
            WHERE id = ?
        `).run(reg.farm_hash, reg.hostname, reg.ip_address || null, reg.mac_address || null, reg.gpu_count, reg.os_version || null, reg.node_id);
    }
    else {
        d.prepare(`
            INSERT INTO nodes (id, farm_hash, hostname, ip_address, mac_address, gpu_count, os_version, status, last_seen_at)
            VALUES (?, ?, ?, ?, ?, ?, ?, 'online', datetime('now'))
        `).run(reg.node_id, reg.farm_hash, reg.hostname, reg.ip_address || null, reg.mac_address || null, reg.gpu_count, reg.os_version || null);
    }
    return d.prepare('SELECT * FROM nodes WHERE id = ?').get(reg.node_id);
}
function getNode(nodeId) {
    const d = getDb();
    const node = d.prepare('SELECT * FROM nodes WHERE id = ?').get(nodeId);
    if (!node)
        return null;
    const latestStat = d.prepare('SELECT payload FROM stats WHERE node_id = ? ORDER BY timestamp DESC LIMIT 1').get(nodeId);
    return {
        ...node,
        latest_stats: latestStat ? JSON.parse(latestStat.payload) : null,
    };
}
function getAllNodes() {
    const d = getDb();
    const nodes = d.prepare('SELECT * FROM nodes ORDER BY last_seen_at DESC').all();
    return nodes.map(node => {
        const latestStat = d.prepare('SELECT payload FROM stats WHERE node_id = ? ORDER BY timestamp DESC LIMIT 1').get(node.id);
        return {
            ...node,
            latest_stats: latestStat ? JSON.parse(latestStat.payload) : null,
        };
    });
}
function getNodesByFarm(farmHash) {
    const d = getDb();
    const nodes = d.prepare('SELECT * FROM nodes WHERE farm_hash = ? ORDER BY hostname').all(farmHash);
    return nodes.map(node => {
        const latestStat = d.prepare('SELECT payload FROM stats WHERE node_id = ? ORDER BY timestamp DESC LIMIT 1').get(node.id);
        return {
            ...node,
            latest_stats: latestStat ? JSON.parse(latestStat.payload) : null,
        };
    });
}
function deleteNode(nodeId) {
    const d = getDb();
    const result = d.prepare('DELETE FROM nodes WHERE id = ?').run(nodeId);
    return result.changes > 0;
}
function updateNodeStatus(nodeId, status) {
    const d = getDb();
    d.prepare('UPDATE nodes SET status = ? WHERE id = ?').run(status, nodeId);
}
/**
 * Mark nodes as offline if they haven't reported stats in `thresholdSecs` seconds.
 */
function markStaleNodes(thresholdSecs = 60) {
    const d = getDb();
    const cutoff = new Date(Date.now() - thresholdSecs * 1000).toISOString().replace('T', ' ').slice(0, 19);
    const stale = d.prepare(`
        SELECT id FROM nodes
        WHERE status = 'online'
        AND last_seen_at < ?
    `).all(cutoff);
    if (stale.length > 0) {
        const stmt = d.prepare("UPDATE nodes SET status = 'offline' WHERE id = ?");
        for (const node of stale) {
            stmt.run(node.id);
        }
        return stale.map(n => n.id);
    }
    return [];
}
// =============================================================================
// Stats Operations
// =============================================================================
function insertStats(nodeId, payload) {
    const d = getDb();
    d.prepare(`
        INSERT INTO stats (node_id, payload, gpu_count, cpu_usage_pct, ram_used_mb, ram_total_mb, toks_per_sec)
        VALUES (?, ?, ?, ?, ?, ?, ?)
    `).run(nodeId, JSON.stringify(payload), payload.gpu_count, payload.cpu.usage_pct, payload.ram.used_mb, payload.ram.total_mb, payload.toks_per_sec);
    // Update node last_seen and status
    d.prepare(`
        UPDATE nodes SET last_seen_at = datetime('now'), status = 'online', gpu_count = ?
        WHERE id = ?
    `).run(payload.gpu_count, nodeId);
}
function getStatsHistory(nodeId, limit = 100) {
    const d = getDb();
    const rows = d.prepare('SELECT payload FROM stats WHERE node_id = ? ORDER BY timestamp DESC LIMIT ?').all(nodeId, limit);
    return rows.map(r => JSON.parse(r.payload));
}
/**
 * Prune stats older than `days` days to keep the DB lean.
 */
function pruneStats(days = 7) {
    const d = getDb();
    const cutoff = new Date(Date.now() - days * 86400000).toISOString().replace('T', ' ').slice(0, 19);
    const result = d.prepare('DELETE FROM stats WHERE timestamp < ?').run(cutoff);
    return result.changes;
}
// =============================================================================
// Command Operations
// =============================================================================
function queueCommand(nodeId, action, params) {
    const d = getDb();
    const id = generateId();
    const payload = params ? JSON.stringify(params) : null;
    d.prepare(`
        INSERT INTO commands (id, node_id, action, payload, status)
        VALUES (?, ?, ?, ?, 'pending')
    `).run(id, nodeId, action, payload);
    return {
        id,
        action,
        ...(params || {}),
    };
}
function getPendingCommands(nodeId) {
    const d = getDb();
    const rows = d.prepare("SELECT * FROM commands WHERE node_id = ? AND status = 'pending' ORDER BY created_at").all(nodeId);
    // Mark them as sent
    const markSent = d.prepare("UPDATE commands SET status = 'sent', sent_at = datetime('now') WHERE id = ?");
    for (const row of rows) {
        markSent.run(row.id);
    }
    return rows.map(row => {
        const params = row.payload ? JSON.parse(row.payload) : {};
        return {
            id: row.id,
            action: row.action,
            ...params,
        };
    });
}
function completeCommand(commandId) {
    const d = getDb();
    d.prepare("UPDATE commands SET status = 'completed', completed_at = datetime('now') WHERE id = ?").run(commandId);
}
// =============================================================================
// Flight Sheet Operations
// =============================================================================
function createFlightSheet(name, description, targets) {
    const d = getDb();
    const id = generateId();
    d.prepare(`
        INSERT INTO flight_sheets (id, name, description, targets)
        VALUES (?, ?, ?, ?)
    `).run(id, name, description, JSON.stringify(targets));
    const row = d.prepare('SELECT * FROM flight_sheets WHERE id = ?').get(id);
    return { ...row, targets: JSON.parse(row.targets) };
}
function getAllFlightSheets() {
    const d = getDb();
    const rows = d.prepare('SELECT * FROM flight_sheets ORDER BY created_at DESC').all();
    return rows.map(r => ({ ...r, targets: JSON.parse(r.targets) }));
}
function getFlightSheet(id) {
    const d = getDb();
    const row = d.prepare('SELECT * FROM flight_sheets WHERE id = ?').get(id);
    if (!row)
        return null;
    return { ...row, targets: JSON.parse(row.targets) };
}
function deleteFlightSheet(id) {
    const d = getDb();
    const result = d.prepare('DELETE FROM flight_sheets WHERE id = ?').run(id);
    return result.changes > 0;
}
/**
 * Apply a flight sheet: queue install_model commands for each target node.
 */
function applyFlightSheet(id) {
    const sheet = getFlightSheet(id);
    if (!sheet)
        return [];
    const commands = [];
    const allNodes = getAllNodes();
    for (const target of sheet.targets) {
        const targetNodes = target.node_id === '*'
            ? allNodes
            : allNodes.filter(n => n.id === target.node_id);
        for (const node of targetNodes) {
            const cmd = queueCommand(node.id, 'install_model', {
                model: target.model,
                gpu: target.gpu,
            });
            commands.push(cmd);
        }
    }
    return commands;
}
function createAlert(nodeId, severity, type, message, value, threshold) {
    const d = getDb();
    const id = generateId();
    d.prepare(`
        INSERT INTO alerts (id, node_id, severity, type, message, value, threshold)
        VALUES (?, ?, ?, ?, ?, ?, ?)
    `).run(id, nodeId, severity, type, message, value, threshold);
    return d.prepare('SELECT * FROM alerts WHERE id = ?').get(id);
}
function getRecentAlerts(limit = 50) {
    const d = getDb();
    return d.prepare('SELECT * FROM alerts ORDER BY created_at DESC LIMIT ?').all(limit);
}
function acknowledgeAlert(alertId) {
    const d = getDb();
    const result = d.prepare('UPDATE alerts SET acknowledged = 1 WHERE id = ?').run(alertId);
    return result.changes > 0;
}
function checkAndAlert(nodeId, stats) {
    const alerts = [];
    // GPU temperature checks
    for (const gpu of stats.gpus) {
        if (gpu.temperatureC > 85) {
            alerts.push(createAlert(nodeId, 'critical', 'gpu_overheat', `GPU overheating: ${gpu.name} at ${gpu.temperatureC}°C`, gpu.temperatureC, 85));
        }
        else if (gpu.temperatureC > 75) {
            alerts.push(createAlert(nodeId, 'warning', 'gpu_hot', `GPU running hot: ${gpu.name} at ${gpu.temperatureC}°C`, gpu.temperatureC, 75));
        }
        // VRAM usage check
        const vramPct = (gpu.vramUsedMb / gpu.vramTotalMb) * 100;
        if (vramPct > 95) {
            alerts.push(createAlert(nodeId, 'warning', 'vram_full', `VRAM nearly full: ${gpu.name} at ${vramPct.toFixed(1)}%`, vramPct, 95));
        }
    }
    // CPU usage check
    if (stats.cpu.usage_pct > 95) {
        alerts.push(createAlert(nodeId, 'warning', 'cpu_saturated', `CPU saturated at ${stats.cpu.usage_pct}%`, stats.cpu.usage_pct, 95));
    }
    // RAM usage check
    const ramPct = (stats.ram.used_mb / stats.ram.total_mb) * 100;
    if (ramPct > 90) {
        alerts.push(createAlert(nodeId, 'warning', 'ram_pressure', `RAM pressure at ${ramPct.toFixed(1)}%`, ramPct, 90));
    }
    // Disk usage check
    const diskPct = (stats.disk.used_gb / stats.disk.total_gb) * 100;
    if (diskPct > 90) {
        alerts.push(createAlert(nodeId, 'critical', 'disk_full', `Disk nearly full at ${diskPct.toFixed(1)}%`, diskPct, 90));
    }
    return alerts;
}
function storeBenchmark(nodeId, result) {
    const d = getDb();
    const id = generateId();
    d.prepare(`
        INSERT INTO benchmarks (id, node_id, model, tokens_per_sec, prompt_eval_rate, eval_rate, total_duration_ms)
        VALUES (?, ?, ?, ?, ?, ?, ?)
    `).run(id, nodeId, result.model, result.tokens_per_sec, result.prompt_eval_rate || 0, result.eval_rate || 0, result.total_duration_ms || 0);
    return d.prepare('SELECT * FROM benchmarks WHERE id = ?').get(id);
}
function getNodeBenchmarks(nodeId, limit = 20) {
    const d = getDb();
    return d.prepare('SELECT * FROM benchmarks WHERE node_id = ? ORDER BY created_at DESC LIMIT ?').all(nodeId, limit);
}
function getAllBenchmarks(limit = 50) {
    const d = getDb();
    return d.prepare('SELECT * FROM benchmarks ORDER BY created_at DESC LIMIT ?').all(limit);
}
function getClusterSummary() {
    const nodes = getAllNodes();
    const onlineNodes = nodes.filter(n => n.status === 'online');
    let totalGpus = 0;
    let totalVram = 0;
    let usedVram = 0;
    let totalToks = 0;
    const modelSet = new Set();
    const farmSet = new Set();
    for (const node of nodes) {
        farmSet.add(node.farm_hash);
        if (node.latest_stats) {
            totalGpus += node.latest_stats.gpu_count;
            totalToks += node.latest_stats.toks_per_sec;
            for (const gpu of node.latest_stats.gpus) {
                totalVram += gpu.vramTotalMb;
                usedVram += gpu.vramUsedMb;
            }
            for (const model of node.latest_stats.inference.loaded_models) {
                modelSet.add(model);
            }
        }
    }
    return {
        total_nodes: nodes.length,
        online_nodes: onlineNodes.length,
        offline_nodes: nodes.length - onlineNodes.length,
        total_gpus: totalGpus,
        total_vram_mb: totalVram,
        used_vram_mb: usedVram,
        total_toks_per_sec: totalToks,
        loaded_models: [...modelSet],
        farm_hashes: [...farmSet],
    };
}
function recordNodeEvent(nodeId, event, detail) {
    const d = getDb();
    d.prepare('INSERT INTO node_events (node_id, event, detail) VALUES (?, ?, ?)').run(nodeId, event, detail || null);
}
function getNodeEvents(nodeId, limit = 50) {
    const d = getDb();
    return d.prepare('SELECT * FROM node_events WHERE node_id = ? ORDER BY created_at DESC LIMIT ?').all(nodeId, limit);
}
/**
 * Get a compact stats history for sparklines: last N data points with key fields extracted.
 */
function getCompactHistory(nodeId, limit = 60) {
    const d = getDb();
    const rows = d.prepare('SELECT payload, timestamp FROM stats WHERE node_id = ? ORDER BY timestamp DESC LIMIT ?').all(nodeId, limit);
    // Reverse so oldest first (for chart rendering)
    rows.reverse();
    const timestamps = [];
    const gpuTemps = [];
    const gpuUtils = [];
    const cpuUsage = [];
    const ramPct = [];
    const toksSec = [];
    for (const row of rows) {
        const stats = JSON.parse(row.payload);
        timestamps.push(row.timestamp);
        cpuUsage.push(stats.cpu.usage_pct);
        ramPct.push(stats.ram.total_mb > 0 ? Math.round((stats.ram.used_mb / stats.ram.total_mb) * 100) : 0);
        toksSec.push(stats.toks_per_sec);
        const temps = [];
        const utils = [];
        for (const gpu of stats.gpus) {
            temps.push(gpu.temperatureC);
            utils.push(gpu.utilizationPct);
        }
        gpuTemps.push(temps);
        gpuUtils.push(utils);
    }
    return { timestamps, gpu_temps: gpuTemps, gpu_utils: gpuUtils, cpu_usage: cpuUsage, ram_pct: ramPct, toks_per_sec: toksSec };
}
/**
 * Find the best node to serve inference for a given model.
 * Strategy: among online nodes with the model loaded, pick the one with lowest average GPU utilization.
 */
function findBestNode(model) {
    const nodes = getAllNodes();
    const candidates = [];
    for (const node of nodes) {
        if (node.status !== 'online' || !node.latest_stats)
            continue;
        const hasModel = node.latest_stats.inference.loaded_models.some(m => m === model || m.startsWith(model.split(':')[0]));
        if (!hasModel)
            continue;
        const gpuUtils = node.latest_stats.gpus.map(g => g.utilizationPct);
        const avgUtil = gpuUtils.length > 0 ? gpuUtils.reduce((a, b) => a + b, 0) / gpuUtils.length : 100;
        candidates.push({
            node_id: node.id,
            hostname: node.hostname,
            ip_address: node.ip_address,
            gpu_utilization_avg: avgUtil,
            in_flight_requests: node.latest_stats.inference.in_flight_requests,
        });
    }
    if (candidates.length === 0)
        return null;
    // Sort by in-flight requests first, then GPU utilization
    candidates.sort((a, b) => {
        if (a.in_flight_requests !== b.in_flight_requests)
            return a.in_flight_requests - b.in_flight_requests;
        return a.gpu_utilization_avg - b.gpu_utilization_avg;
    });
    return candidates[0];
}
/**
 * Get all unique models loaded across the cluster.
 */
function getClusterModels() {
    const nodes = getAllNodes();
    const modelMap = new Map();
    for (const node of nodes) {
        if (node.status !== 'online' || !node.latest_stats)
            continue;
        for (const model of node.latest_stats.inference.loaded_models) {
            const list = modelMap.get(model) || [];
            list.push(node.id);
            modelMap.set(model, list);
        }
    }
    return [...modelMap.entries()].map(([model, nodeIds]) => ({
        model,
        node_count: nodeIds.length,
        nodes: nodeIds,
    }));
}
function getHealthScore() {
    const nodes = getAllNodes();
    if (nodes.length === 0) {
        return {
            score: 0, grade: 'F', color: 'red',
            factors: { nodes_online_pct: 0, avg_gpu_temp: 0, avg_vram_headroom_pct: 100, recent_critical_alerts: 0, has_loaded_models: false },
        };
    }
    const onlinePct = (nodes.filter(n => n.status === 'online').length / nodes.length) * 100;
    let totalTemp = 0, tempCount = 0;
    let totalVramPct = 0, vramCount = 0;
    let hasModels = false;
    for (const node of nodes) {
        if (!node.latest_stats)
            continue;
        for (const gpu of node.latest_stats.gpus) {
            totalTemp += gpu.temperatureC;
            tempCount++;
            if (gpu.vramTotalMb > 0) {
                totalVramPct += (gpu.vramUsedMb / gpu.vramTotalMb) * 100;
                vramCount++;
            }
        }
        if (node.latest_stats.inference.loaded_models.length > 0)
            hasModels = true;
    }
    const avgTemp = tempCount > 0 ? totalTemp / tempCount : 0;
    const avgVramUsedPct = vramCount > 0 ? totalVramPct / vramCount : 0;
    const vramHeadroom = 100 - avgVramUsedPct;
    // Count recent critical alerts (last hour)
    const d = getDb();
    const oneHourAgo = new Date(Date.now() - 3600000).toISOString().replace('T', ' ').slice(0, 19);
    const criticalAlerts = d.prepare("SELECT COUNT(*) as cnt FROM alerts WHERE severity = 'critical' AND created_at > ?").get(oneHourAgo).cnt;
    // Calculate score (100 = perfect)
    let score = 0;
    // Nodes online: 0-30 points
    score += Math.min(30, (onlinePct / 100) * 30);
    // GPU temps: 0-25 points (below 60°C = perfect, above 85°C = 0)
    if (tempCount > 0) {
        const tempScore = avgTemp < 60 ? 25 : avgTemp > 85 ? 0 : 25 * (1 - (avgTemp - 60) / 25);
        score += Math.max(0, tempScore);
    }
    else {
        score += 25; // No GPUs = no temp concerns
    }
    // VRAM headroom: 0-20 points (>30% free = perfect, <5% = 0)
    if (vramCount > 0) {
        const vramScore = vramHeadroom > 30 ? 20 : vramHeadroom < 5 ? 0 : 20 * (vramHeadroom / 30);
        score += Math.max(0, vramScore);
    }
    else {
        score += 20;
    }
    // No critical alerts: 0-15 points
    const alertPenalty = Math.min(15, criticalAlerts * 5);
    score += 15 - alertPenalty;
    // Models loaded: 0-10 points
    if (hasModels)
        score += 10;
    score = Math.round(Math.max(0, Math.min(100, score)));
    const grade = score >= 90 ? 'A' : score >= 75 ? 'B' : score >= 60 ? 'C' : score >= 40 ? 'D' : 'F';
    const color = score > 80 ? 'green' : score > 50 ? 'yellow' : 'red';
    return {
        score, grade, color,
        factors: {
            nodes_online_pct: Math.round(onlinePct),
            avg_gpu_temp: Math.round(avgTemp),
            avg_vram_headroom_pct: Math.round(vramHeadroom),
            recent_critical_alerts: criticalAlerts,
            has_loaded_models: hasModels,
        },
    };
}
function createSchedule(name, type, cron, config) {
    const d = getDb();
    const id = generateId();
    const nextRun = computeNextRun(cron);
    d.prepare(`
        INSERT INTO schedules (id, name, type, cron, config, next_run)
        VALUES (?, ?, ?, ?, ?, ?)
    `).run(id, name, type, cron, JSON.stringify(config), nextRun);
    return getSchedule(id);
}
function getSchedule(id) {
    const d = getDb();
    const row = d.prepare('SELECT * FROM schedules WHERE id = ?').get(id);
    if (!row)
        return null;
    return { ...row, config: JSON.parse(row.config), enabled: !!row.enabled };
}
function getAllSchedules() {
    const d = getDb();
    const rows = d.prepare('SELECT * FROM schedules ORDER BY created_at DESC').all();
    return rows.map(r => ({ ...r, config: JSON.parse(r.config), enabled: !!r.enabled }));
}
function deleteSchedule(id) {
    const d = getDb();
    return d.prepare('DELETE FROM schedules WHERE id = ?').run(id).changes > 0;
}
function toggleSchedule(id, enabled) {
    const d = getDb();
    return d.prepare('UPDATE schedules SET enabled = ? WHERE id = ?').run(enabled ? 1 : 0, id).changes > 0;
}
function markScheduleRun(id) {
    const d = getDb();
    const schedule = getSchedule(id);
    if (!schedule)
        return;
    const nextRun = computeNextRun(schedule.cron);
    d.prepare('UPDATE schedules SET last_run = datetime(\'now\'), next_run = ? WHERE id = ?').run(nextRun, id);
}
function getDueSchedules() {
    const d = getDb();
    const now = new Date().toISOString().replace('T', ' ').slice(0, 19);
    const rows = d.prepare('SELECT * FROM schedules WHERE enabled = 1 AND (next_run IS NULL OR next_run <= ?)').all(now);
    return rows.map(r => ({ ...r, config: JSON.parse(r.config), enabled: !!r.enabled }));
}
// =============================================================================
// SSH Key Management
// =============================================================================
function addSshKey(nodeId, label, publicKey) {
    const d = getDb();
    const id = generateId();
    // Compute a simple fingerprint from the key (SHA-256 of the base64 portion)
    const keyParts = publicKey.trim().split(/\s+/);
    const keyData = keyParts.length >= 2 ? keyParts[1] : keyParts[0];
    // Simple hash for fingerprint — just first 16 chars of base64
    const fingerprint = 'SHA256:' + keyData.slice(0, 43).replace(/[+/=]/g, '');
    d.prepare(`
        INSERT INTO ssh_keys (id, node_id, label, public_key, fingerprint)
        VALUES (?, ?, ?, ?, ?)
    `).run(id, nodeId, label, publicKey.trim(), fingerprint);
    return d.prepare('SELECT * FROM ssh_keys WHERE id = ?').get(id);
}
function getNodeSshKeys(nodeId) {
    const d = getDb();
    return d.prepare('SELECT * FROM ssh_keys WHERE node_id = ? ORDER BY created_at DESC').all(nodeId);
}
function deleteSshKey(keyId) {
    const d = getDb();
    return d.prepare('DELETE FROM ssh_keys WHERE id = ?').run(keyId).changes > 0;
}
// =============================================================================
// Node Tags
// =============================================================================
function addNodeTag(nodeId, tag) {
    const d = getDb();
    d.prepare('INSERT OR IGNORE INTO node_tags (node_id, tag) VALUES (?, ?)').run(nodeId, tag.toLowerCase().trim());
}
function removeNodeTag(nodeId, tag) {
    const d = getDb();
    return d.prepare('DELETE FROM node_tags WHERE node_id = ? AND tag = ?').run(nodeId, tag.toLowerCase().trim()).changes > 0;
}
function getNodeTags(nodeId) {
    const d = getDb();
    const rows = d.prepare('SELECT tag FROM node_tags WHERE node_id = ? ORDER BY tag').all(nodeId);
    return rows.map(r => r.tag);
}
function getNodesByTag(tag) {
    const d = getDb();
    const nodeIds = d.prepare('SELECT node_id FROM node_tags WHERE tag = ?').all(tag.toLowerCase().trim());
    return nodeIds.map(r => getNode(r.node_id)).filter((n) => n !== null);
}
function getAllTags() {
    const d = getDb();
    return d.prepare('SELECT tag, COUNT(*) as count FROM node_tags GROUP BY tag ORDER BY count DESC').all();
}
// =============================================================================
// Model Pull Progress
// =============================================================================
function startModelPull(nodeId, model) {
    const d = getDb();
    const id = generateId();
    d.prepare(`
        INSERT INTO model_pulls (id, node_id, model) VALUES (?, ?, ?)
    `).run(id, nodeId, model);
    return d.prepare('SELECT * FROM model_pulls WHERE id = ?').get(id);
}
function updateModelPull(nodeId, model, progress) {
    const d = getDb();
    const sets = ["updated_at = datetime('now')"];
    const vals = [];
    if (progress.status) {
        sets.push('status = ?');
        vals.push(progress.status);
    }
    if (progress.progress_pct !== undefined) {
        sets.push('progress_pct = ?');
        vals.push(progress.progress_pct);
    }
    if (progress.bytes_downloaded !== undefined) {
        sets.push('bytes_downloaded = ?');
        vals.push(progress.bytes_downloaded);
    }
    if (progress.bytes_total !== undefined) {
        sets.push('bytes_total = ?');
        vals.push(progress.bytes_total);
    }
    vals.push(nodeId, model);
    d.prepare(`UPDATE model_pulls SET ${sets.join(', ')} WHERE node_id = ? AND model = ? AND status = 'downloading'`).run(...vals);
}
function getActiveModelPulls(nodeId) {
    const d = getDb();
    return d.prepare("SELECT * FROM model_pulls WHERE node_id = ? AND status IN ('downloading', 'verifying') ORDER BY started_at DESC").all(nodeId);
}
function getAllActiveModelPulls() {
    const d = getDb();
    return d.prepare("SELECT * FROM model_pulls WHERE status IN ('downloading', 'verifying') ORDER BY started_at DESC").all();
}
/**
 * Simple cron-to-next-run computation.
 * Supports: @hourly, @daily, @every Nh, @every Nm
 */
function computeNextRun(cron) {
    const now = new Date();
    let next;
    if (cron === '@hourly') {
        next = new Date(now.getTime() + 3600000);
    }
    else if (cron === '@daily') {
        next = new Date(now.getTime() + 86400000);
    }
    else if (cron.startsWith('@every ')) {
        const match = cron.match(/@every\s+(\d+)([hms])/);
        if (match) {
            const val = parseInt(match[1]);
            const unit = match[2];
            const ms = unit === 'h' ? val * 3600000 : unit === 'm' ? val * 60000 : val * 1000;
            next = new Date(now.getTime() + ms);
        }
        else {
            next = new Date(now.getTime() + 3600000); // default 1h
        }
    }
    else {
        // Default: 1 hour
        next = new Date(now.getTime() + 3600000);
    }
    return next.toISOString().replace('T', ' ').slice(0, 19);
}
